apps/gcl-create.8
.Dd August 2, 2015
.Dt GCL-CREATE 8
.Os Swarm-GDP
.Sh NAME
.Nm gcl-create
.Nd create new GDP log
.Sh SYNOPSIS
.Nm gcl-create
.Op Fl C Ar creator-name
.Op Fl D Ar debug-spec
.Op Fl e Ar key-crypto
.Op Fl G Ar router-ip
.Op Fl h Ar hash-alg
.Op Fl k Ar key-type
.Op Fl K Ar key-location
.Op Fl b Ar key-bits
.Op Fl c Ar curve-name
.Op Fl q
.Op Fl s Ar logd-gdp-addr
.Op Fl S
.Op Ar mdid = Ar metadata
.Op Ar external-name
.Sh DESCRIPTION
.Nm gcl-create
creates a new GDP log on a given log server.
In its default mode it also creates a public/secret keypair
associated with this log, with the public key stored in the log metadata.
Additional (text-based) metadata can be added
(arguably this should allow binary metadata as well).
Optionally the log can be given a human-understandable name;
if not specified a seemingly random name is made up.
.Ss "Metadata"
User-specified metadata may be added when a log is created.
The metadata name is stored as a 32-bit number and is generally represented
by four characters.
.Pp
In addition to user metadata,
.Nm gcl-create
adds metadata for the log creation time,
the public half of the signing key in DER format,
and the human-readable name.
.Pp
Metadata is immutable; there is no way to add, delete, or change metadata
after the log is created.
.Ss "Warning"
This implementation will change in the future,
and the command will probably be subsumed into a swarm service.
.Sh OPTIONS
.Bl -tag
.It Fl C
Sets the name of the creator in the log metadata.
If this is not specified it will be derived from system data.
It should be in the form of an email address.
.It Fl D Ar debug-spec
Turns on debugging flags using the libep-style format.
Useful only with the code in hand.
.It Fl e Ar key-crypto
Specifies symmetric encryption algorithm to use for storing the secret key.
If
.Ar key-crypto
is
.Li none
then the key is stored unencrypted on disk.
Defaults to
.Va swarm.gdp.crypto.keyenc.alg .
.It Fl G Ar router-ip
Contact the given IP address (which may be a DNS name)
to get a connection to the GDP routing layer.
This will be replaced by automatic discovery
(for example, using Zeroconf)
soon.
The address may have an optional port number after a colon.
If this is not specified,
the names listed in the
.Va swarm.gdp.routers
runtime administrative parameter
are searched.
.It Fl h Ar hash-alg
Selects the hash (message digest) algorithm used when
generating and checking signatures.
When data is appended to the log,
the log name, log metadata, record number, and record data
is hashed using this algorithm and then that hash is signed.
.It Fl k Ar key-type
Selects the type of signing key to be created for this log.
It can be one of
.Li rsa ,
.Li dsa ,
or
.Li ec .
As a special case, the type
.Li none
suppresses creation of the keypair.
.It Fl K Ar key-location
This option has two purposes.
First, if you want to use an existing keypair
(formatted as a PEM file)
you can use the
.Fl K
flag to name the existing PEM file.
If you want
.Nm
to create a keypair for you,
.Fl K
can name either a file in which to store the secret key,
or a directory into which the secret key will be written
named the same as the internal name of the log (with
.Va \&.pem
appended).
If
.Fl K
is not specified, the value of the
.Va swarm.gdp.crypto.key.dir
administrative parameter will be used as a directory name,
which defaults to
.Pa ./KEYS ,
or to the current directory if the
.Pa KEYS
directory does not exist.
.It Fl b Ar key-bits
If the key type is RSA or DSA,
this parameter specifies the number of bits in the key.
If this is omitted and the key type is RSA,
this defaults to the value of
.Va swarm.gdp.crypto.rsa.keylen .
DSA keys use
.Va swarm.gdp.crypto.dsa.keylen .
Additionally, RSA keys use the
.Va swarm.gdp.crypto.rsa.keyexp
parameter to set the exponent.
This parameter is ignored when creating EC key pairs.
.It Fl c Ar curve-name
Sets the name of the curve to use for Elliptic Cryptography keys.
This overrides the
.Va swarm.gdp.crypto.ec.curve
parameter.
Ignored when creating RSA or DSA key pairs.
.It Fl q
If set,
.Nm
does not show errors or informational message.
In particular, using
.Fl q
is a way to ensure that a log exists without give an error if it already exists.
.
.It Fl s Ar logd-gdp-addr
The name of the log daemon on which to create the physical instantiation
of the log.
If not specified, the value of the
.Va swarm.gdp.gcl-create.server
administrative will be used.
If that is not set,
it will default to one of the servers at Berkeley.
If you want a specific server at your site,
either use the
.Fl s
flag or set that administrative parameter.
.
.It Fl S
Skips the test to see if the log already exists before creating it.
Used for debugging.
.El
.Sh EXIT STATUS
.Bl -tag
.It Li EX_OK No (0)
Success
.It Li EX_CANTCREAT No (73)
The log already exists.
.It Li EX_DATAERR No (65)
A command-line key file could not be parsed.
.It Li EX_IOERR No (74)
A secret key file could not be created.
.It Li EX_NOHOST No (68)
The log server name was not valid.
.It Li EX_NOINPUT No (67)
A command-line key file could not be opened.
.It Li EX_SOFTWARE No (70)
A successfully created public key could not be converted to external (DER) form
for storage in new log metadata,
or severe internal error.
.It Li EX_USAGE No (64)
Command line parameters are incorrect.
.It Li EX_UNAVAILABLE No (69)
A specified key length was insecure.
It was impossible to create a new key.
Other unspecified error.
.El
.Sh ADMINISTRATIVE PARAMETERS
.Bl -tag
.It swarm.gdp.crypto.dsa.keylen
The DSA key length.
Defaults to 2048.
Overridden by
.Fl b .
.It swarm.gdp.crypto.ec.curve
The EC curve.
Defaults to
.Li sect283r1
(also known as
.Li B-283 ) .
Overridden by
.Fl c .
.It swarm.gdp.crypto.hash.alg
The hash algorithm.
Defaults to
.Li sha256 .
Overridden by
.Fl h .
.It swarm.gdp.crypto.key.dir
The directory in which to store secret keys.
Defaults to
.Pa KEYS
(in the current working directory).
If that subdirectory does not exist,
the keys are written to the current directory.
Overridden by
.Fl K .
.It swarm.gdp.crypto.keyenc.alg
The secret key (symmetric) encryption algorithm.
Defaults to
.Li aes192 .
Overridden by
.Fl e .
.It swarm.gdp.crypto.rsa.keyexp
The exponent to be used in the RSA algorithm.
Defaults to 3.
.It swarm.gdp.crypto.rsa.keylen
The key length for the RSA algorithm.
Defaults to 2048.
Overridden by
.Fl b .
.It swarm.gdp.crypto.sign.alg
The signing algorithm.
Defaults to
.Li ec .
Overridden by
.Fl k .
.It swarm.gdp.gcl-create.server
The default server on which to create logs.
Overridden by
.Fl s .
.El
.\".Sh ENVIRONMENT
.\".Sh FILES
.Sh SEE ALSO
.Xr gdp 7 ,
.Xr gdplogd 8
.Sh EXAMPLES
.Bl -bullet
.It
To create a GCL on a default server:
.Dl gcl-create edu.berkeley.cs.eric.log00
The secret key will be written into a a file named
.Pa KEYS/<something>.pem ,
where
.Va <something>
is the base-64-encoded internal name of the log, e.g.,
.Li 1KZy5jy1QpghTe8QBmDQGqdz3a_9tVP3qp6uxlOeJdk.pem .
The password for encrypting this key will be read from the standard input.
.It
To create a GCL on log server named edu.berkeley.eecs.gdp-01 using defaults:
.Dl logd=edu.berkeley.eecs.gdp-01.gdplogd
.Dl gcl-create -s $logd edu.berkeley.cs.eric.log01
.It
To create a log named
.Li $newlog
on the server named
.Li $logd
encrypted with a 1024-bit RSA key,
leaving the unencrypted key in a file named
.Pa mykey.pem :
.Dl gcl-create -k RSA -b 1024 -e none -K mykey.pem $newlog
.It
To create a log with user-specified metadata:
.Dl gcl-create Qo "MYMD=My special metadata" Qc $newlog
.It
To create a log without a human-friendly name using sha-224
as the hash (message digest) algorithm:
.Dl gcl-create -h sha224
.Nm gcl-create
will print the base-64-encoded name of the new log.
You should be careful to record that,
since it is your only way to access the log.
.El
.Sh BUGS
There should be some way to provide the key password
other than reading it from the standard input.
.Pp
It should not be necessary to name the server on which the log will be created.
This command will be replaced by a smart service
that chooses placement based on locality and capacity.


apps/gcl-create.c
/* vim: set ai sw=4 sts=4 ts=4 : */

/*
**  LOG-CREATE --- create a new log
**
**		This is a temporary interface.  Ultimately we need a log
**		creation service that does reasonable log placement rather
**		than having to name a specific log server.
**
**	----- BEGIN LICENSE BLOCK -----
**	Applications for the Global Data Plane
**	From the Ubiquitous Swarm Lab, 490 Cory Hall, U.C. Berkeley.
**
**	Copyright (c) 2015-2017, Regents of the University of California.
**	All rights reserved.
**
**	Permission is hereby granted, without written agreement and without
**	license or royalty fees, to use, copy, modify, and distribute this
**	software and its documentation for any purpose, provided that the above
**	copyright notice and the following two paragraphs appear in all copies
**	of this software.
**
**	IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
**	SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
**	PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
**	EVEN IF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
**
**	REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
**	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
**	FOR A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION,
**	IF ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO
**	OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
**	OR MODIFICATIONS.
**	----- END LICENSE BLOCK -----
*/

#include <ep/ep.h>
#include <ep/ep_app.h>
#include <ep/ep_crypto.h>
#include <ep/ep_dbg.h>
#include <ep/ep_string.h>
#include <gdp/gdp.h>

#include <openssl/rsa.h>
#include <openssl/engine.h>
#include <openssl/pem.h>

#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <netdb.h>
#include <pwd.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>
#include <sys/stat.h>


// minimum secure key length
#ifndef GDP_MIN_KEY_LEN
# define GDP_MIN_KEY_LEN		1024
#endif // GDP_MIN_KEY_LEN

static EP_DBG	Dbg = EP_DBG_INIT("gdp.gcl-create", "Create new log app");


const char *
select_logd_name(void)
{
	const char *p;

	p = ep_adm_getstrparam("swarm.gdp.gcl-create.server", NULL);
	if (p == NULL)
	{
		// seed the random number generator
		struct timeval tv;
		gettimeofday(&tv, NULL);
		srandom(tv.tv_usec);

		// default to one of the berkeley servers
		int r = random() % 4 + 1;
		int l = sizeof "edu.berkeley.eecs.gdp-00.gdplogd" + 1;
		char *nbuf;

		nbuf = ep_mem_malloc(l);
		snprintf(nbuf, l, "edu.berkeley.eecs.gdp-0%d.gdplogd", r);
		p = nbuf;
	}

	ep_dbg_cprintf(Dbg, 3, "Using log server %s\n", p);

	return p;
}

// command line options
#define OPTIONS		"b:c:C:D:e:G:h:k:K:qs:S"

void
usage(void)
{
	fprintf(stderr, "Usage: %s [-C creator] [-D dbgspec] [-e key_enc_alg]\n"
			"\t[-G gdpd_addr] [-q] [-s logd_name] [-S]\n"
			"\t[-h] [-k keytype] [-K keyfile] [-b keybits] [-c curve]\n"
			"\t[<mdid>=<metadata>...] [gcl_name]\n"
			"    -C  set name of log creator (owner; for metadata)\n"
			"    -D  set debugging flags\n"
			"    -e  set secret key encryption algorithm\n"
			"    -G  IP host to contact for GDP router\n"
			"    -q  don't give error if log already exists\n"
			"    -s  set name of log daemon on which to create the log\n"
			"    -S  skip the test to see if log already exists (for debugging)\n"
			"    -h  set the hash (message digest) algorithm (defaults to sha256)\n"
			"    -k  type of key; valid key types are \"rsa\", \"dsa\", and \"ec\"\n"
			"\t(defaults to ec); \"none\" turns off key generation\n"
			"    -K  use indicated public key/place to write secret key\n"
			"\tIf -K specifies a directory, a .pem file is written there\n"
			"\twith the name of the GCL (defaults to \"KEYS\" or \".\")\n"
			"    -b  set size of key in bits (RSA and DSA only)\n"
			"    -c  set curve name (EC only)\n"
			"    creator is the id of the creator, formatted as an email address\n"
			"    logd_name is the name of the log server to host this log\n"
			"    metadata ids are (by convention) four letters or digits\n"
			"    gcl_name is the name of the log to create\n",
			ep_app_getprogname());
	exit(EX_USAGE);
}


int
main(int argc, char **argv)
{
	gdp_name_t gcliname;			// internal name of GCL
	const char *gclxname = NULL;	// external name of GCL
	gdp_name_t logdiname;			// internal name of log daemon
	const char *logdxname = NULL;	// external name of log daemon
	const char *cname = NULL;		// creator/owner name (for metadata)
	gdp_gcl_t *gcl = NULL;
	gdp_gclmd_t *gmd = NULL;
	int opt;
	EP_STAT estat;
	char *gdpd_addr = NULL;
	bool show_usage = false;
	bool make_new_key = true;
	bool quiet = false;
	bool skip_existence_test = false;
	int md_alg_id = -1;
	int keytype = EP_CRYPTO_KEYTYPE_UNKNOWN;
	int keylen = 0;
	int exponent = 0;
	const char *curve = NULL;
	const char *keyfile = NULL;
	int keyform = EP_CRYPTO_KEYFORM_PEM;
	int key_enc_alg = -1;
	char *passwd = NULL;
	EP_CRYPTO_KEY *key = NULL;
	char *p;

	// quick pass so debugging is on for initialization
	while ((opt = getopt(argc, argv, OPTIONS)) > 0)
	{
		if (opt == 'D')
			ep_dbg_set(optarg);
	}
	optind = 1;
#if EP_OSCF_NEED_OPTRESET
	optreset = 1;
#endif

	// preinit library (must be early due to crypto code in arg processing)
	gdp_lib_init(NULL);

	// collect command-line arguments
	while ((opt = getopt(argc, argv, OPTIONS)) > 0)
	{
		switch (opt)
		{
		 case 'b':
			keylen = atoi(optarg);
			break;

		 case 'c':
			curve = optarg;
			break;

		 case 'C':
			cname = optarg;
			break;

		 case 'D':
			// already done
			break;

		 case 'e':
			key_enc_alg = ep_crypto_keyenc_byname(optarg);
			if (key_enc_alg < 0)
			{
				ep_app_error("unknown key encryption algorithm %s", optarg);
				show_usage = true;
			}
			break;

		 case 'G':
			gdpd_addr = optarg;
			break;

		 case 'h':
			md_alg_id = ep_crypto_md_alg_byname(optarg);
			if (md_alg_id < 0)
			{
				ep_app_error("unknown digest hash algorithm %s", optarg);
				show_usage = true;
			}
			break;

		 case 'k':
			if (strcasecmp(optarg, "none") == 0)
			{
				make_new_key = false;
				break;
			}
			keytype = ep_crypto_keytype_byname(optarg);
			if (keytype == EP_CRYPTO_KEYTYPE_UNKNOWN)
			{
				ep_app_error("unknown key type %s", optarg);
				show_usage = true;
			}
			break;

		 case 'K':
			keyfile = optarg;
			break;

		 case 'q':
			quiet = true;
			break;

		 case 's':
			logdxname = optarg;
			break;

		 case 'S':
			skip_existence_test = true;
			break;

		 default:
			show_usage = true;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (show_usage || argc < 0)
		usage();

	// collect any metadata
	gmd = gdp_gclmd_new(0);
	while (argc > 0 && (p = strchr(argv[0], '=')) != NULL)
	{
		gdp_gclmd_id_t mdid = 0;
		int i;

		p++;
		for (i = 0; i < 4; i++)
		{
			if (argv[0][i] == '=')
				break;
			mdid = (mdid << 8) | (unsigned) argv[0][i];
		}

		gdp_gclmd_add(gmd, mdid, strlen(p), p);

		argc--;
		argv++;
	}

	// back compatibility for old calling convention
	if (argc == 2 && logdxname == NULL)
	{
		logdxname = *argv++;
		argc--;
	}

	// if we don't have a log daemon, pick one
	if (logdxname == NULL)
		logdxname = select_logd_name();

	gdp_parse_name(logdxname, logdiname);

	// name is optional ; if omitted one will be created
	if (argc-- > 0)
		gclxname = *argv++;

	if (show_usage || argc > 0)
		usage();

	// initialize the GDP library
	estat = gdp_init(gdpd_addr);
	if (!EP_STAT_ISOK(estat))
	{
		ep_app_error("GDP Initialization failed");
		goto fail0;
	}

	// allow thread to settle to avoid interspersed debug output
	ep_time_nanosleep(INT64_C(100000000));

	if (gclxname != NULL)
	{
		gdp_parse_name(gclxname, gcliname);
		if (!skip_existence_test)
		{
			// make sure it doesn't already exist
			estat = gdp_gcl_open(gcliname, GDP_MODE_RO, NULL, &gcl);
			if (EP_STAT_ISOK(estat))
			{
				// oops, we shouldn't be able to open it
				(void) gdp_gcl_close(gcl);
				if (!quiet)
					ep_app_error("Cannot create %s: already exists", gclxname);
				exit(EX_CANTCREAT);
			}
		}
	}

	/**************************************************************
	**  Set up other automatic metadata
	*/

	// creation time
	{
		EP_TIME_SPEC tv;
		char timestring[40];

		ep_time_now(&tv);
		ep_time_format(&tv, timestring, sizeof timestring, EP_TIME_FMT_DEFAULT);
		gdp_gclmd_add(gmd, GDP_GCLMD_CTIME, strlen(timestring), timestring);
	}

	// creator id
	{
		char cnamebuf[200];

		if (cname == NULL)
		{
			/*
			**  Find the name of the creator.
			*/

			// user name (from password file)
			char *uname;
			char unamebuf[40];
			struct passwd *pw = getpwuid(getuid());

			if (pw != NULL)
			{
				uname = pw->pw_name;
			}
			else
			{
				snprintf(unamebuf, sizeof unamebuf, "%d", getuid());
				uname = unamebuf;
			}

			// fully qualified domain name
			//  (Linux sets HOST_NAME_MAX too low, so we use a magic constant)
			char fqdn[1025];

			// gethostname doesn't guarantee null termination
			fqdn[sizeof fqdn - 1] = '\0';
			if (gethostname(fqdn, sizeof fqdn - 1) != 0)
			{
				ep_app_error("Cannot find current host name");
				strlcpy(fqdn, "localhost", sizeof fqdn);
			}

			if (strchr(fqdn, '.') == NULL)
			{
				// need to tack on a domain name
				struct addrinfo hints, *ai;
				int i;

				memset(&hints, 0, sizeof hints);
				hints.ai_family = AF_UNSPEC;
				hints.ai_socktype = SOCK_STREAM;
				hints.ai_flags = AI_CANONNAME;

				if ((i = getaddrinfo(fqdn, NULL, &hints, &ai)) != 0)
				{
					ep_app_error("Cannot find DNS domain name: %s",
							gai_strerror(i));
				}
				else
				{
					strlcpy(fqdn, ai->ai_canonname, sizeof fqdn);
				}
			}

			// now put it together
			snprintf(cnamebuf, sizeof cnamebuf, "%s@%s", uname, fqdn);
			cname = cnamebuf;
		}
		ep_dbg_cprintf(Dbg, 1, "Creating log as %s\n", cname);
		gdp_gclmd_add(gmd, GDP_GCLMD_CID, strlen(cname), cname);
	}


	/**************************************************************
	**  Do cryptographic setup (e.g., generating keys)
	*/

	if (key_enc_alg < 0)
	{
		const char *p = ep_adm_getstrparam("swarm.gdp.crypto.keyenc.alg",
								"aes192");
		key_enc_alg = ep_crypto_keyenc_byname(p);
		if (key_enc_alg < 0)
		{
			ep_app_error("unknown secret key encryption algorithm %s", p);
			exit(EX_USAGE);
		}
	}

	if (md_alg_id < 0)
	{
		const char *p = ep_adm_getstrparam("swarm.gdp.crypto.hash.alg",
								"sha256");
		md_alg_id = ep_crypto_md_alg_byname(p);
		if (md_alg_id < 0)
		{
			ep_app_error("unknown digest hash algorithm %s", p);
			exit(EX_USAGE);
		}
	}
	if (keytype == EP_CRYPTO_KEYTYPE_UNKNOWN)
	{
		const char *p = ep_adm_getstrparam("swarm.gdp.crypto.sign.alg", "ec");
		keytype = ep_crypto_keytype_byname(p);
		if (keytype == EP_CRYPTO_KEYTYPE_UNKNOWN)
		{
			ep_app_error("unknown keytype %s", p);
			exit(EX_USAGE);
		}
	}
	switch (keytype)
	{
	case EP_CRYPTO_KEYTYPE_RSA:
		if (keylen <= 0)
			keylen = ep_adm_getintparam("swarm.gdp.crypto.rsa.keylen", 2048);
		exponent = ep_adm_getintparam("swarm.gdp.crypto.rsa.keyexp", 3);
		break;

	case EP_CRYPTO_KEYTYPE_DSA:
		if (keylen <= 0)
			keylen = ep_adm_getintparam("swarm.gdp.crypto.dsa.keylen", 2048);
		break;

	case EP_CRYPTO_KEYTYPE_EC:
		if (curve == NULL)
			curve = ep_adm_getstrparam("swarm.gdp.crypto.ec.curve", NULL);
		break;
	}

	// see if we have an existing key
	bool keyfile_is_directory = false;
	if (keyfile == NULL)
	{
		// nope -- create a key in current directory if requested
		if (make_new_key)
		{
			struct stat st;

			keyfile_is_directory = true;
			keyfile = ep_adm_getstrparam("swarm.gdp.crypto.key.dir", "KEYS");
			if (stat(keyfile, &st) != 0 || (st.st_mode & S_IFMT) != S_IFDIR)
				keyfile = ".";
		}
	}
	else
	{
		// might be a directory
		struct stat st;
		FILE *key_fp;

		if (stat(keyfile, &st) == 0 && (st.st_mode & S_IFMT) == S_IFDIR)
		{
			keyfile_is_directory = true;
		}
		else if ((key_fp = fopen(keyfile, "r")) == NULL)
		{
			// no existing key file; OK if we are creating it
			if (!make_new_key)
			{
				ep_app_error("Could not open secret key %s", keyfile);
				exit(EX_NOINPUT);
			}
		}
		else
		{
			key = ep_crypto_key_read_fp(key_fp, keyfile,
						keyform, EP_CRYPTO_F_SECRET);
			if (key == NULL)
			{
				ep_app_error("Could not read secret key %s", keyfile);
				exit(EX_DATAERR);
			}
			else
			{
				make_new_key = false;
			}
			fclose(key_fp);
		}
	}

	// if creating new key, go ahead and do it (in memory)
	if (make_new_key)
	{
		if (keylen < GDP_MIN_KEY_LEN && keytype != EP_CRYPTO_KEYTYPE_EC)
		{
			ep_app_error("Insecure key length %d; %d min",
					keylen, GDP_MIN_KEY_LEN);
			exit(EX_UNAVAILABLE);
		}
		key = ep_crypto_key_create(keytype, keylen, exponent, curve);
		if (key == NULL)
		{
			ep_app_error("Could not create new key");
			exit(EX_UNAVAILABLE);
		}
	}

	/*
	**	Now would be a good time to write the secret key to disk
	**		We do this before creating GCL so we don't risk having
	**		GCLs that exist with no access possible.  But we need
	**		to use a temporary name because we don't know the final
	**		name of the GCL yet.
	*/

	char *tempkeyfile = NULL;
	if (make_new_key)
	{
		const char *localkeyfile = keyfile;

		// it might be a directory, in which case we use a temporary name
		if (keyfile_is_directory)
		{
			size_t len;
			gdp_pname_t pbuf;
			gdp_name_t tempname;

			// use random name for the temp file
			evutil_secure_rng_get_bytes(tempname, sizeof tempname);
			gdp_printable_name(tempname, pbuf);
			len = strlen(keyfile) + sizeof pbuf + 6;
			tempkeyfile = ep_mem_malloc(len);
			snprintf(tempkeyfile, len, "%s/%s.pem", keyfile, pbuf);
			localkeyfile = tempkeyfile;
		}

		int fd;
		FILE *fp;

		if ((fd = open(localkeyfile, O_WRONLY|O_CREAT|O_TRUNC, 0600)) < 0 ||
				(fp = fdopen(fd, "w")) == NULL)
		{
			ep_app_error("Cannot create %s", localkeyfile);
			exit(EX_IOERR);
		}

		estat = ep_crypto_key_write_fp(key, fp, EP_CRYPTO_KEYFORM_PEM,
						key_enc_alg, passwd, EP_CRYPTO_F_SECRET);

		// TODO: should really clear the fp buffer memory here to
		//		 avoid exposing any secret key information after free
		fclose(fp);

		if (!EP_STAT_ISOK(estat))
		{
			char ebuf[200];

			// remove the zero length key file
			unlink(localkeyfile);

			ep_app_abort("Couldn't write secret key to %s: %s",
					localkeyfile, ep_stat_tostr(estat, ebuf, sizeof ebuf));
		}
	}

	// at this point, key is initialized and secret is on disk
	if (key != NULL)
	{
		// add the public key to the metadata
		uint8_t der_buf[EP_CRYPTO_MAX_DER + 4];
		uint8_t *derp = der_buf + 4;

		der_buf[0] = md_alg_id;
		der_buf[1] = keytype;
		der_buf[2] = (keylen >> 8) & 0xff;
		der_buf[3] = keylen & 0xff;
		estat = ep_crypto_key_write_mem(key, derp, EP_CRYPTO_MAX_DER,
						EP_CRYPTO_KEYFORM_DER, key_enc_alg, NULL,
						EP_CRYPTO_F_PUBLIC);
		if (!EP_STAT_ISOK(estat))
		{
			ep_app_error("Could not create DER format public key");
			exit(EX_SOFTWARE);
		}

		gdp_gclmd_add(gmd, GDP_GCLMD_PUBKEY,
				EP_STAT_TO_INT(estat) + 4, der_buf);
	}

	/*
	**  Hello sailor, this is where the actual creation happens
	*/

	gdp_parse_name(logdxname, logdiname);
	if (gclxname == NULL)
	{
		// create a new GCL handle with a new name based on metadata
		estat = gdp_gcl_create(NULL, logdiname, gmd, &gcl);
	}
	else
	{
		// save the external name as metadata
		if (gmd == NULL)
			gmd = gdp_gclmd_new(0);
		gdp_gclmd_add(gmd, GDP_GCLMD_XID, strlen(gclxname), gclxname);

		// create a GCL with the provided name
		estat = gdp_gcl_create(gcliname, logdiname, gmd, &gcl);
	}
	EP_STAT_CHECK(estat, goto fail1);

	/*
	**  Rename the secret key file to the new name.
	*/

	if (tempkeyfile != NULL)
	{
		char *finalkeyfile = NULL;
		size_t len;
		gdp_pname_t pbuf;

		gdp_printable_name(*gdp_gcl_getname(gcl), pbuf);
		len = strlen(keyfile) + sizeof pbuf + 6;
		finalkeyfile = ep_mem_malloc(len);
		snprintf(finalkeyfile, len, "%s/%s.pem", keyfile, pbuf);
		if (rename(tempkeyfile, finalkeyfile) != 0)
		{
			ep_app_abort("Cannot rename %s to %s", tempkeyfile, finalkeyfile);
		}

		if (!quiet)
		{
			ep_app_info("New secret key is in %s", finalkeyfile);
			ep_app_info("Save this file!  You'll need it to write the new log");
		}

		ep_mem_free(finalkeyfile);
		ep_mem_free(tempkeyfile);
		tempkeyfile = NULL;
	}

	// just for a lark, let the user know the (internal) name
	if (!quiet)
	{
		gdp_pname_t pname;

		printf("Created new GCL %s\n\ton log server %s\n",
				gdp_printable_name(*gdp_gcl_getname(gcl), pname), logdxname);
	}

	gdp_gcl_close(gcl);

fail1:
	// free metadata, if set
	if (gmd != NULL)
		gdp_gclmd_free(gmd);
	// if tempkeyfile did not get consumed (renamed), remove it
	if (tempkeyfile != NULL)
	{
		remove(tempkeyfile);
		tempkeyfile = NULL;
	}

fail0:
	;				// avoid compiler error
	int exitstat;

	if (EP_STAT_ISOK(estat))
		exitstat = EX_OK;
	else if (EP_STAT_IS_SAME(estat, GDP_STAT_NAK_NOROUTE))
		exitstat = EX_NOHOST;
	else if (EP_STAT_ISABORT(estat))
		exitstat = EX_SOFTWARE;
	else
		exitstat = EX_UNAVAILABLE;

	// OK status can have values; hide that from the user
	if (EP_STAT_ISOK(estat))
		estat = EP_STAT_OK;

	if (!EP_STAT_ISOK(estat))
		ep_app_message(estat, "exiting with status");
	else if (!quiet)
		fprintf(stderr, "Exiting with status OK\n");

	return exitstat;
}
